### PR TITLE
NXP-21968: Don't index root document in elasticsearch

### PR DESCRIPTION
**Pull request by kevinleturc**
T&P: https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-kleturc-2/29

### CODE REVIEW

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java [hunk 1]
```diff
@@ -106,10 +106,12 @@ public void beforeCompletion() {
 
     @Override
     public void afterCompletion(int status) {
-        if (getAllCommands().isEmpty()) {
-            return;
-        }
+        // un hook the current listener even if there's no commands to index
```
Comment (kevinleturc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java]: 👍

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java [hunk 2]
```diff
@@ -106,10 +106,12 @@ public void beforeCompletion() {
...
+        // unless, during next transaction this listener won't be hooked to it
```
Comment (bdelbosc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java]: not sure to understand the "unless" part and why it is related to the indexing of root document

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestAutomaticIndexing.java [hunk 3]
```diff
@@ -61,6 +61,7 @@
 import org.nuxeo.ecm.core.api.security.ACL;
 import org.nuxeo.ecm.core.api.security.ACP;
 import org.nuxeo.ecm.core.api.security.SecurityConstants;
+import org.nuxeo.ecm.core.api.security.impl.ACLImpl;
```
Comment (bdelbosc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestAutomaticIndexing.java]: dup

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestAutomaticIndexing.java [hunk 4]
```diff
@@ -176,6 +177,35 @@ public void setupIndex() throws Exception {
         esa.initIndexes(true);
     }
 
+//    @Test
+    public void shouldNotIndexRootDocument() throws Exception {
+        startTransaction();
+        // update acp on root document as it is reinit during tear down
```
Comment (bdelbosc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestAutomaticIndexing.java]: could be interesting to add some assertion that root is never indexed
- after a framework init
- after a reindex all
- after a reindex with nxql using the root uid

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java [hunk 5]
```diff
@@ -128,7 +128,7 @@ public void iCanPerformESQLPageProviderOnRepository() throws IOException, Interr
 
         JsonNode node = mapper.readTree(response.getEntityInputStream());
         // Verify results
-        assertEquals(19, getLogEntries(node).size());
+        assertEquals(20, getLogEntries(node).size());
```
Comment (bdelbosc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java]: weird that if we are not indexing the root document anymore we have one more entry ?

Comment (kevinleturc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java]: Indeed, previously we get a result node with 19 items, but page size is 20 and the total size of query was 21. I didn't succeed to reproduce it after the fix?

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java (hunk 2 above)
Comment (kevinleturc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/listener/ElasticSearchInlineListener.java]: Now we update document root during cleanupSession. Which triggers an UPDATE event, if we don't index the root document, there'll be no commands to index but the listener is still hooked to the current transaction and do not un-hook. During next test, we still use the same thread, listener will have `isEnlisted.get() == true` and it won't be hooked to current transaction.

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java (hunk 5 above)
Comment (efge) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java]: So are we confident we know what happens here? It's worth getting to the bottom of this to be sure we don't have a hidden bug somewhere else.

**Comment by efge** [comment]
👍

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java (hunk 5 above)
Comment (kevinleturc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/rest/RestESDocumentsTest.java]: We're not. I'll merge this PR and try to find the issue after it.

**Pull request merged**